Reinstate product groups

WalterKolczynski-NOAA opened the pull request:
# Description
Testing with full-sized GEFS found that the sheer number of tasks overloads rocoto, resulting in `rocotorun` taking over 10 min to complete or hanging entirely. To reduce the number of tasks, product groups are reimplemented so that multiple forecast hour are processed in a single task. However, the implementation is a little different than previously.

The jobs where groups are enabled (atmos_products, oceanice_products, wavepostsbs, atmos_ensstat, and gempak) have a new variable, `MAX_TASKS`, that controls how many groups to use. This setting is currently *per member*. The forecast hours to be processed are then divided into this many groups as evenly as possible without crossing forecast segment boundaries. The walltime for those jobs is then multiplied by the number of times in the largest group. For the gridded wave post job, the dependencies were also updated to trigger off of either the data being available or the appropriate segment completing (the dependencies had not been updated when the job was initially broken into fhrs).

A number of helper methods are added to Tasks to determine these groups and make a standard metatask variable dict in a centralized location. There is also a function to multiply the walltime, but this may be better off relocated to wxflow with the other time functions.

As part of switching from a single value to a list, hours are no longer passed by rocoto as zero-padded values. The lists are comma-delimited (without spaces) and split apart in the job stub (`jobs/rocoto/*`), so each j-job call is still a single forecast hour.

The offline post (upp) job is not broken into groups, since it really isn't used outside the analysis anymore.
Resolves #2999
Resolves #3210

# Type of change
- [x] Bug fix (fixes something broken)
- [x] New feature (adds functionality)
- [x] Maintenance (code refactor, clean-up, new CI test, etc.)

# Change characteristics
- Is this a breaking change (a change in existing functionality)? NO
- Does this change require a documentation update? NO
- Does this change require an update to any of the following submodules? NO (If YES, please add a link to any PRs that are pending.)
  - [ ] EMC verif-global <!-- NOAA-EMC/EMC_verif-global#1234 -->
  - [ ] GDAS <!-- NOAA-EMC/GDASApp#1234 -->
  - [ ] GFS-utils <!-- NOAA-EMC/gfs-utils#1234 -->
  - [ ] GSI <!-- NOAA-EMC/GSI#1234 -->
  - [ ] GSI-monitor <!-- NOAA-EMC/GSI-Monitor#1234 -->
  - [ ] GSI-utils <!-- NOAA-EMC/GSI-Utils#1234 -->
  - [ ] UFS-utils <!-- ufs-community/UFS_UTILS#1234 -->
  - [ ] UFS-weather-model <!-- ufs-community/ufs-weather-model#1234 -->
  - [ ] wxflow <!-- NOAA-EMC/wxflow#1234 -->

# How has this been tested?
- Full-sized GEFS test on Hercules
- Standard forecast-only test on Hercules
- Standard atm-only cycled test on Hercules

# Checklist
- [x] Any dependent changes have been merged and published
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have documented my code, including function, input, and output descriptions
- [x] My changes generate no new warnings
- [x] New and existing tests pass with my changes
- [x] This change is covered by an existing CI test or a new one has been added
- [x] Any new scripts have been added to the .github/CODEOWNERS file with owners
- [x] I have made corresponding changes to the system documentation if necessary

WalterKolczynski-NOAA committed:
Reinstate product groups

Testing with full-sized GEFS found that the sheer number of tasks
overloads rocoto, resulting in `rocotorun` taking over 10 min to
complete or hanging entirely. To reduce the number of tasks, product
groups are reimplemented so that multiple forecast hour are processed
in a single task. However, the implementation is a little different
than previously.

The jobs where groups are enabled (atmos_products, oceanice_products,
and wavepostsbs) have a new variable, `MAX_TASKS`, that controls how
many groups to use. This setting is currently *per member*. The
forecast hours to be processed are then divided into this many groups
as evenly as possible without crossing forecast segment boundaries.
The walltime for those jobs is then multiplied by the number of times
in the largest group.

A number of helper methods are added to Tasks to determine these
groups and make a standard metatask variable dict in a centralized
location. There is also a function to multiply the walltime, but this
may be better off relocated to wxflow with the other time functions.

As part of switching from a single value to a list, hours are no longer
passed by rocoto as zero-padded values. The lists are comma-delimited
(without spaces) and split apart in the job stub (`jobs/rocoto/*`), so
each j-job call is still a single forecast hour.

The offline post (upp) job is not broken into groups, since it really
isn't used outside the analysis anymore. Gempak jobs that run over
multiple forecast hours also aren't broken into groups yet.

Resolves #2999

KateFriedman-NOAA commented on workflow/rocoto/tasks.py:
```diff
@@ -176,6 +176,203 @@ def _get_forecast_hours(run, config, component='atmos') -> List[str]:
+
+        Returns
+        -------
+        str: String represnting a time delta in HH:MM:SS format
```
Small typo fix:
```suggestion
                       String representing a time delta in HH:MM:SS format
        multiplier: int | float
                    Value to multiply the time delta by

        Returns
        -------
        str: String representing a time delta in HH:MM:SS format
```

KateFriedman-NOAA commented on workflow/rocoto/tasks.py:
```diff
@@ -176,6 +176,203 @@ def _get_forecast_hours(run, config, component='atmos') -> List[str]:
+        -------
+        dict: Several variable lists for use in rocoto metatasks:
+              fhr_list: list of comma-separated lists of fhr groups
+              fhr_label: list of labels corrsponding to the fhr range
```
Small typo fix:
```suggestion
              fhr_label: list of labels corresponding to the fhr range
```

KateFriedman-NOAA commented on workflow/rocoto/tasks.py:
```diff
@@ -176,6 +176,203 @@ def _get_forecast_hours(run, config, component='atmos') -> List[str]:
+                       {'fhrs': [18, 19, 20, 21, 22, 23], 'seg': 1}]
+        assert Tasks.get_job_groups(fhrs=test_array, ngroups=4, breakpoints=[11]) == test_groups
+
+        # Test with a break point not at a normal splilt point
```
Small typo fix:
```suggestion
        # Test with a break point not at a normal split point
```

aerorahul commented on workflow/rocoto/gefs_tasks.py:
```diff
@@ -308,22 +310,35 @@ def atmos_ensstat(self):
 
+        fhrs = self._get_forecast_hours('gefs', self._configs['wavepostsbs'], 'wave')
+        is_replay = self._configs['wavepostsbs']['REPLAY_ICS']
+        if is_replay:
+            fhrs = [fhr for fhr in fhrs if fhr not in [0, 1, 2]]
```
Can you place a comment on what this is doing and why?

WalterKolczynski-NOAA replied:
Done

aerorahul commented on workflow/rocoto/tasks.py:
```diff
@@ -176,6 +176,203 @@ def _get_forecast_hours(run, config, component='atmos') -> List[str]:
+        return groups
+
+    @staticmethod
+    def test_job_groups():
```
Should we create and place the tests in `test_tasks.py`?  How are these tests executed?

WalterKolczynski-NOAA replied:
Right now, they are only executed in an interactive python interpreter by importing the class and then calling the test methods.

Instinctually I like the tests in the same file they are testing, but after doing some cursory research, it seems the preferred way is a separate file, so should probably go in that direction. I can either do it here, or in a future PR.

aerorahul replied:
Since tests were added (a good thing), I suggest placing them in the separate file such as `test_tasks.py`.  
It will take a little effort to make them part of a `pytest`, but that can be addressed in another PR.

aerorahul reviewed (commented):
I have no major concerns with this PR.  It looks good and clean to me.  I have not created the XML and looked at it or run any tests.

A few comments inline with the review.

KateFriedman-NOAA reviewed (changes requested):
Changes look good, just leaving a few suggested changes to resolve some typos I noticed.

aerorahul commented on workflow/rocoto/gfs_tasks.py:
```diff
@@ -1194,13 +1197,21 @@ def wavepostsbs(self):
         deps.append(rocoto.add_dependency(dep_dict))
         dependencies = rocoto.create_dependency(dep=deps)
```
Lets update the dependency here to launch the forecast hour (group) post-processing sbs job when data is available.  This will resolve #3210

WalterKolczynski-NOAA committed:
Merge branch 'develop' into feature/prod_groups

EricSinsky-NOAA commented on jobs/rocoto/atmos_products.sh:
```diff
@@ -15,13 +15,17 @@ if (( status != 0 )); then exit "${status}"; fi
+	###############################################################
+	# Execute the JJOB
+	###############################################################
+	"${HOMEgfs}/jobs/JGLOBAL_ATMOS_PRODUCTS"
+	status=$?
```
I recommend moving `jobid` here and naming it to something like the suggestion below. Please see my comment above. 
```suggestion
  jobid="${job}_${FORECAST_HOUR}.$$"
	"${HOMEgfs}/jobs/JGLOBAL_ATMOS_PRODUCTS"
	status=$?
```

aerorahul replied:
excellent suggestion @EricSinsky-NOAA

WalterKolczynski-NOAA replied:
Added these for all the groupified jobs.

EricSinsky-NOAA commented on jobs/rocoto/atmos_products.sh:
```diff
@@ -15,13 +15,17 @@ if (( status != 0 )); then exit "${status}"; fi
 export job="atmos_products"
 export jobid="${job}.$$"
```
If `KEEPDATA` is `YES`, the `DATA` (`DATA=${DATA:-"${DATAROOT}/${jobid}"}`) of the current `FORECAST_HOUR` may overwrite the `DATA` of the previous `FORECAST_HOUR`, which may potentially create issues. To avoid these potential issues, particularly when `KEEPDATA` is needed for debugging, can `jobid` be moved inside the for `FORECAST_HOUR` loop? I think this suggestion also can be applied to `oceanice_products.sh` and `wavepostsbs.sh`. 
```suggestion
```

aerorahul commented on jobs/rocoto/oceanice_products.sh:
```diff
@@ -15,11 +15,17 @@ if (( status != 0 )); then exit "${status}"; fi
+	###############################################################
+	# Execute the JJOB
+	###############################################################
+	"${HOMEgfs}/jobs/JGLOBAL_OCEANICE_PRODUCTS"
```
@EricSinsky-NOAA's suggestion for jobid should be done here as well.  In addition, since rocoto sets `COMPONENT`, may be we can use that as well in `jobid` such that `${job}_${COMPONENT}_${FORECAST_HOUR}.$$`
```suggestion
  jobid="${job}_${COMPONENT}_${FORECAST_HOUR}.$$"
	"${HOMEgfs}/jobs/JGLOBAL_OCEANICE_PRODUCTS"
```

aerorahul commented on jobs/rocoto/wavepostsbs.sh:
```diff
@@ -5,17 +5,24 @@ source "${HOMEgfs}/ush/preamble.sh"
+for FORECAST_HOUR in "${fhr_list[@]}"; do
+	FHR3=$(printf '%03d' "${FORECAST_HOUR}")
+	# Execute the JJOB
+	"${HOMEgfs}/jobs/JGLOBAL_WAVE_POST_SBS"
```
```suggestion
  jobid="${job}_${FORECAST_HOUR}.$$"
	"${HOMEgfs}/jobs/JGLOBAL_WAVE_POST_SBS"
```

WalterKolczynski-NOAA committed:
Add wave gridded post data dependency

Adds an alternative data dependency for gridded wave post so jobs can
begin before the entire forecast is complete. Uses the next file
existing to prevent reading incomplete files.

Resolves #3210

WalterKolczynski-NOAA committed:
Fix documentation typos in tasks.py

WalterKolczynski-NOAA committed:
Use fhr in jobid for product jobs

WalterKolczynski-NOAA committed:
Add missing whitespace

WalterKolczynski-NOAA committed:
Add comment about removing output hours for replay

WalterKolczynski-NOAA commented:
All reviewer comments other than the location of test methods have been addressed, so I am going to go ahead and run CI for the first machine.

aerorahul committed:
Merge branch 'develop' into feature/prod_groups

WalterKolczynski-NOAA committed:
Move new task tests to separate file

Moves the new task tests to a separate file. Also changes the permissions
on the existing test files so they can be executed.

WalterKolczynski-NOAA committed:
Merge branch 'develop' into feature/prod_groups

EricSinsky-NOAA commented on workflow/rocoto/gefs_tasks.py:
```diff
@@ -308,22 +310,43 @@ def atmos_ensstat(self):
         return task
 
     def wavepostsbs(self):
+
```
In addition to wavepostsbs, should the atmos_ensstat tasks be grouped as well?

WalterKolczynski-NOAA replied:
Yes, also the dependency there is wrong anyway and I need to fix. Not sure how that passed.

WalterKolczynski-NOAA replied:
Changes made and a test is running. It passed the first time because I ran the default case instead of the full-sized one. The default case has fewer output times than the max tasks, so grouping never kicked in.

WalterKolczynski-NOAA replied:
First couple ensstat groups have now run successfully. Will keep test running, but can proceed now.

WalterKolczynski-NOAA committed:
Switch from unittest to pytest for Tasks test

WalterKolczynski-NOAA committed:
Extend grouping to ensstat

WalterKolczynski-NOAA committed:
Reset permissions on workflow tests

WalterKolczynski-NOAA committed:
Move workflow tests to tests directory

WalterKolczynski-NOAA committed:
Update ensstat rocoto stub to loop over fhr

WalterKolczynski-NOAA committed:
Replace tabs with spaces in tasks test

How did those tabs get in there?

WalterKolczynski-NOAA committed:
Fix ensstat exit code

aerorahul reviewed (dismissed):
looks good.  thanks for including the suggestions.

WalterKolczynski-NOAA committed:
Switch type hint from pipe to Union for backwards compatability

WalterKolczynski-NOAA committed:
Use bisect instead of index to split fhrs

Switches from using index to using bisect_array. This makes the
split more forgiving, since the fhr of the breakpoint need not be
in the fhr array.

Unit tests still pass with this change.

WalterKolczynski-NOAA committed:
Remove empty arrays after breakpoint splitting

EricSinsky-NOAA commented on parm/config/gefs/config.resources:
```diff
@@ -242,14 +243,15 @@ case ${step} in
     ;;
 
   "atmos_ensstat")
```
```suggestion
  "atmos_ensstat")
   # Walltime is per forecast hour; will be multipled by group size
```

WalterKolczynski-NOAA replied:
Added. Thanks for catching that.

EricSinsky-NOAA reviewed (dismissed):
Changes overall look good to me. I just included one minor optional suggestion. Thank you for addressing my previous suggestions.

WalterKolczynski-NOAA committed:
Break atmos gempak into groups

WalterKolczynski-NOAA committed:
Add comment about walltime for ensstat

WalterKolczynski-NOAA committed:
Fix copy/paste error

WalterKolczynski-NOAA commented:
Need to look at the gempak jobs a little closer.

aerorahul committed:
Merge branch 'develop' into feature/prod_groups

WalterKolczynski-NOAA committed:
Fix gempak stub

WalterKolczynski-NOAA committed:
Merge branch 'develop' into feature/prod_groups

KateFriedman-NOAA reviewed (approved):
Looks good, thanks @WalterKolczynski-NOAA !
